### PR TITLE
Stress test scripts to launch in GKE

### PR DESCRIPTION
**Pull request by sreecha**
NOTE: This is the same one as  https://github.com/grpc/grpc/pull/5394 which I accidentally closed. Sorry about that.

**Description**
- `test/cpp/interop/metrics_client.cc` :  Added timeout for the APIs
- ` test/cpp/util/metrics_server.cc `: Log level correction
- `tools/big_query/big_query_utils.py`: Helper functions to interact with BigQuery 
- `tools/dockerfile/grpc_interop_stress_cxx/Dockerfile:` Install google-api-python-client in the image for BigQuery APIs (since the stress client/server now write status to BigQuery)
- `tools/dockerfile/grpc_interop_stress_cxx/build_interop_stress.sh`:  Package metrics client as well in the base docker image for stress tests
- `tools/gke/kubernetes_api.py` : Minor cosmetic changes
-  `tools/gke/run_stress_tests_on_gke.py`: This is the main script that will orchestrate everything. 
- `tools/run_tests/stress_test/run_client.py`: This is a wrapper python scripts *that runs inside the stress client pod* is launches the actual stress client program. The wrapper performs other housekeeping tasks like writing status to bigquery, querying metrics and uploading to BigQuery and more importantly NOT exiting in case of an error (this way, we get a chance to connect to the pod and examine error logs etc)
- `tools/run_tests/stress_test/run_server.py`: The wrapper program for stress server. Similar to run_client.py
- `tools/run_tests/stress_test/stress_test_utils.py`: Common utils used by `run_client.py` and `run_server.py`

**Remaining work**
- Setup a jenkins worker to call `run_stress_tests_on_gke.py` once a day

### CODE REVIEW

File: tools/big_query/big_query_utils.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python2.7
```
Comment (jtattermusch) [inline tools/big_query/big_query_utils.py]: nit: not sure we want an dedicated directory for something as specific as big_query utils.
tools/stresstest/.... maybe?

File: tools/gke/kubernetes_api.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env python2.7
```
Comment (jtattermusch) [inline tools/gke/kubernetes_api.py]: maybe also move under tools/stresstest?

File: tools/gke/kubernetes_api.py [hunk 3]
```diff
@@ -214,3 +223,47 @@ def delete_pod(kube_host, kube_port, namespace, pod_name):
...
+  if not is_success:
+    print 'Error in creating Service'
+    return False
+
+  print 'Successfully created the pod/service %s' % pod_name
+  return True
+
+
+def delete_pod_and_service(kube_host, kube_port, namespace, pod_name):
+  """ A simple helper function that calls delete_pod and delete_service """
```
Comment (jtattermusch) [inline tools/gke/kubernetes_api.py]: nit: s/simple//

File: tools/gke/run_stress_tests_on_gke.py [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+  cmd = ['tools/jenkins/build_interop_stress_image.sh']
+  p = subprocess.Popen(args=cmd)
+  retcode = p.wait()
+  if retcode != 0:
+    print 'Error in building docker image'
+    return False
+
+  print 'Adding an additional tag %s to the image %s' % (tag_name, image_name)
+  cmd = ['docker', 'tag', '-f', image_name, tag_name]
+  p = subprocess.Popen(args=cmd)
```
Comment (jtattermusch) [inline tools/gke/run_stress_tests_on_gke.py]: you could also do subprocess.call()
https://github.com/grpc/grpc/blob/master/tools/run_tests/dockerjob.py#L92

Comment (jtattermusch) [inline tools/gke/run_stress_tests_on_gke.py]: also, can you tag  the resulting image at the end of build_interop_stress_image?
That would save you ~20lines of code here.

File: tools/gke/run_stress_tests_on_gke.py [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+    return False
+
+  return True
+
+
+def _push_docker_image_to_gke_registry(docker_tag_name):
+  """Executes 'gcloud docker push <docker_tag_name>' to push the image to GKE registry"""
+  cmd = ['gcloud', 'docker', 'push', docker_tag_name]
+  print 'Pushing %s to GKE registry..' % docker_tag_name
+  p = subprocess.Popen(args=cmd)
```
Comment (jtattermusch) [inline tools/gke/run_stress_tests_on_gke.py]: maybe subprocess.call()?

File: tools/run_tests/stress_test/run_client.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+  # Update status that the test is starting (in the status table)
+  bq_helper.insert_summary_row(EventType.STARTING, details)
+
+  metrics_cmd = [metrics_client_image
+                ] + [x for x in metrics_client_args_str.split()]
+  stress_cmd = [image_name] + [x for x in args_str.split()]
+
+  print 'Launching process %s ...' % stress_cmd
+  stress_p = subprocess.Popen(args=stress_cmd,
```
Comment (jtattermusch) [inline tools/run_tests/stress_test/run_client.py]: wouldn't you benefit from reusing some code from tools/run_tests/jobset.py ?

File: tools/gke/kubernetes_api.py (hunk 2 above)
Comment (sreecha) [inline tools/gke/kubernetes_api.py]: both big_query_utils.py and kubernetes_api.py are actually very generic and can be used by other tests as well. Do you recommend I move them both under tools/cloud" ?

Comment (jtattermusch) [inline tools/gke/kubernetes_api.py]: tools/cloud or tools/gcp works fine

File: tools/gke/kubernetes_api.py (hunk 3 above)
Comment (sreecha) [inline tools/gke/kubernetes_api.py]: Done

File: tools/gke/run_stress_tests_on_gke.py (hunk 4 above)
Comment (sreecha) [inline tools/gke/run_stress_tests_on_gke.py]: Done

File: tools/gke/run_stress_tests_on_gke.py (hunk 5 above)
Comment (sreecha) [inline tools/gke/run_stress_tests_on_gke.py]: Done.

File: tools/run_tests/stress_test/run_client.py (hunk 6 above)
Comment (sreecha) [inline tools/run_tests/stress_test/run_client.py]: Not by much. There are a few things that jobset doesn't give me (which I could have added to jobset) like for example: I want the output from stress_test on stdout (or to a well-known/configurable file path - while jobset writes stdout to a temp file)..I need to read the results from metrics_client's stdout (i.e pipe the results back) and parse them before uploading to bigquery - jobset doesn't have this ... etc..

Just seemed easier to directly call Subprocess.Popen ..

**Comment by sreecha** [comment]
I have addressed all the comments and pushed the latest changes.
Thanks

**Comment by jtattermusch** [comment]
Santity test complains about some missing copyrights.
test/cpp/interop/metrics_client.cc modified 2016 but copyright only extends to 2015
copyright missing: tools/gcp/utils/big_query_utils.py
copyright missing: tools/gcp/utils/kubernetes_api.py

Fine to merge after fixing the copyrights.

**Pull request merged**